Python/ast.c

```diff
@@ -326,12 +326,234 @@ validate_expr(expr_ty exp, expr_context_ty ctx)
 }
 
 static int
-validate_pattern(expr_ty p)
+validate_pattern_attribute(expr_ty pattern)
 {
-    // Coming soon (thanks Batuhan)!
+    if (!validate_expr(pattern, Load)) {
+        return 0;
+    }
+
+    switch (pattern->kind) {
+        case Attribute_kind:
+            return validate_pattern_attribute(pattern->v.Attribute.value);
+        case Name_kind:
+            return 1;
+        default:
+            PyErr_SetString(PyExc_ValueError,
+                            "invalid node type for an attribute pattern");
+            return 0;
+    }
+}
+
+static int
+validate_pattern(expr_ty pattern)
+{
+    PyObject *literal;
+    int op, left, right;
+    Py_ssize_t i, length;
+    asdl_expr_seq *values, *keys;
+    switch (pattern->kind) {
+        case Attribute_kind:
+            return validate_pattern_attribute(pattern);
+        case MatchOr_kind:
+            values = pattern->v.MatchOr.patterns;
+            length = asdl_seq_LEN(values);
+            if (length < 2) {
+                PyErr_SetString(PyExc_ValueError,
+                                "MatchOr expects at least 2 patterns");
+                return 0;
+            }
+            for (i = 0; i < length; i++) {
+                if (!validate_pattern(asdl_seq_GET(values, i))) {
+                    return 0;
+                }
+            }
+            return 1;
+        case Call_kind:
+            if (!validate_pattern_attribute(pattern->v.Call.func)) {
+                return 0;
+            }
+
+            values = pattern->v.Call.args;
+            length = asdl_seq_LEN(values);
+
+            expr_ty arg;
+            for (i = 0; i < length; i++) {
+                arg = asdl_seq_GET(values, i);
+                if (arg->kind == Starred_kind) {
+                    PyErr_SetString(PyExc_ValueError,
+                                    "can't use single star pattern inside of "
+                                    "a Call pattern");
+                    return 0;
+                }
+                if (!validate_pattern(arg)) {
+                    return 0;
+                }
+            }
+
+            asdl_keyword_seq *keywords = pattern->v.Call.keywords;
+            length = asdl_seq_LEN(keywords);
+            keyword_ty keyword;
+            for (i = 0; i < length; i++) {
+                keyword = asdl_seq_GET(keywords, i);
+                if (keyword->arg == Py_None) {
```

Comment:
I believe this will be `NULL` for `**kwargs`, not `None`:

```suggestion
                if (!keyword->arg) {
```